// spectacle/es/components/presenter-components.js
import styled from 'react-emotion';
export var ContentContainer =
/*#__PURE__*/
styled("div", {
  target: "e1c6hj4s0"
})("display:flex;flex:1;padding:10px 50px 0;");
export var EndHeader =
/*#__PURE__*/
styled("h1", {
  target: "e1c6hj4s1"
})("color:#fff;display:flex;margin:0;");
export var PresenterContent =
/*#__PURE__*/
styled("div", {
  target: "e1c6hj4s2"
})("display:flex;flex:1;flex-direction:column;height:100%;width:100%;");
export var PreviewPane =
/*#__PURE__*/
styled("div", {
  target: "e1c6hj4s3"
})("display:flex;flex:1;flex-wrap:wrap;height:90%;justify-content:center;position:absolute;top:10%;width:60%;");
export var PreviewCurrentSlide =
/*#__PURE__*/
styled("div", {
  target: "e1c6hj4s4"
})("border:2px #fff solid;display:flex;flex:0 0 100%;height:55%;width:55%;padding:20px;");
export var PreviewNextSlide =
/*#__PURE__*/
styled("div", {
  target: "e1c6hj4s5"
})("align-items:center;display:flex;flex:0 0 68.75%;height:40%;justify-content:center;opacity:0.4;");
export var HeaderContainer =
/*#__PURE__*/
styled("div", {
  target: "e1c6hj4s6"
})("position:absolute;display:block;color:#fff;width:100%;height:10%;text-align:center;padding:10px 50px;");
export var Notes =
/*#__PURE__*/
styled("div", {
  target: "e1c6hj4s7"
})("color:#fff;display:block;height:90%;left:calc(60% + 50px);padding:10px 30px;position:absolute;top:10%;width:(40% - 100px);overflow:auto;");
export var SlideInfo =
/*#__PURE__*/
styled("h2", {
  target: "e1c6hj4s8"
})("position:relative;top:50%;transform:translateY(-50%);float:left;margin:0;line-height:1;display:inline-block;fontSize:28px;");
